src/backend/distributed/operations/shard_rebalancer.c

```diff
@@ -1609,6 +1646,88 @@ RebalanceTableShards(RebalanceOptions *options, Oid shardReplicationModeOid)
 }
 
 
+/*
+ * RebalanceTableShardsBackground rebalances the shards for the relations
+ * inside the relationIdList across the different workers. It does so using our
+ * background job+task infrastructure.
+ */
+static void
+RebalanceTableShardsBackground(RebalanceOptions *options, Oid shardReplicationModeOid)
+{
+	char transferMode = LookupShardTransferMode(shardReplicationModeOid);
+	EnsureReferenceTablesExistOnAllNodesExtended(transferMode);
+
+	if (list_length(options->relationIdList) == 0)
+	{
+		return;
+	}
+
+	Oid relationId = InvalidOid;
+	char *operationName = "rebalance";
+	if (options->drainOnly)
+	{
+		operationName = "move";
+	}
+
+	foreach_oid(relationId, options->relationIdList)
+	{
+		AcquireRebalanceColocationLock(relationId, operationName);
+	}
+
+	List *placementUpdateList = GetRebalanceSteps(options);
+
+	if (list_length(placementUpdateList) == 0)
+	{
+		return;
+	}
+
+	/* find the name of the shard transfer mode to interpolate in the scheduled command */
+	Datum shardTranferModeLabelDatum =
+		DirectFunctionCall1(enum_out, shardReplicationModeOid);
+	char *shardTranferModeLabel = DatumGetCString(shardTranferModeLabelDatum);
+
+	if (list_length(placementUpdateList) == 0)
+	{
+		return;
+	}
+
+	/* schedule planned moves */
+	int64 jobId = CreateBackgroundJob("rebalance", "Rebalance colocation group ...");
+
+	PlacementUpdateEvent *move = NULL;
+	StringInfoData buf = { 0 };
+	initStringInfo(&buf);
+	bool first = true;
+	int64 prevJobId = 0;
+	foreach_ptr(move, placementUpdateList)
+	{
+		resetStringInfo(&buf);
+
+		appendStringInfo(&buf,
+						 "SELECT citus_move_shard_placement(%ld,%s,%u,%s,%u,%s)",
```

Comment:
Would be good to prefix the function call with pg_catalog to prevent overloading attempts

ideally, we would also make the operation idempotent

Reply:
opened a PR for the idempotency part https://github.com/citusdata/citus/pull/6313/files